### DISCUSSION TITLE
Which is the preferred resample for Decklink out?

### DISCUSSION THREAD

**Discussion by alatteri** [CESNET/UltraGrid#265] [Q&A] [answered]
Hi,

What is the preferred resampler for fixing the Decklink out audio issue?   I see that recently there was a check-in for parallel Speex.

Thanks,
Alan

**Comment by alatteri** [comment]
In my testing, speex produces audible clicks in the output, and they are not reported as Buffer Overflows.  soxr does not produce these audible errors, so I will stick with that.

**Comment by MartinPulec** [comment, marked as answer]
Thanks for the info, this agrees with the [info](https://github.com/CESNET/UltraGrid/pull/225) from from guys at Sohonet, I believe that this is caused when changed resample rate, because Speex resampler is destroyed&recreated while SoxR allows changing the rate on fly, keeping its internal state. So if there occurs sample rate change, it does produce the artifacts.

So I've pushed [a change](https://github.com/CESNET/UltraGrid/commit/c5395fe662be170270474b2ca035ad011703c44d) that implicitly select SoxR for DeckLink drift fixer.

I don't think speex is bad in general, if using it for "static" resample, the above doesn't occur. The true is that SoxR presumably has better quality (from the [article](https://lastique.github.io/src_test/), I've found). On the other hand, Speex doesn't add a delay. From the performance point of view I think it is a tie – SoxR should be be better but it is used as single threaded. Speex is multi-threaded in UG.
